stdlib/result.gr

```diff
@@ -1,89 +1,201 @@
-    match (v) {
-        Err(e) => fn(e),
-        Ok(x)  => Ok(x)
-    }
+/**
+ * If the Result is an `Err`, applies the given function to the `Err` value and returns the result. Returns the unmodified `Ok` value otherwise.
+ *
+ * @param fn: The function to apply
+ * @param result: The result to map
+ * @returns `fn(errValue)` if the result is of the `Err` variant, errValue represent the value wrapped by the result, otherwise returns the unmodified `Ok`
+ */
+export let flatMapErr = (fn, result) => {
+  match (result) {
+    Err(e) => fn(e),
+    Ok(x) => Ok(x),
+  }
 }
 
-export let map = (fn, v) => {
-    match (v) {
-        Ok(x)  => Ok(fn(x)),
-        Err(e) => Err(e)
-    }
+/**
+ * If the Result is `Ok(x)`, returns `Ok(fn(x))`. Returns the unmodified `Err` otherwise.
+ *
+ * @param fn: The function to map
+ * @param result: The result to map on
+ * @returns `Ok(fn(okValue))` if the result is of the `Ok` variant, otherwise returns the unmodified `Err`
+ */
+export let map = (fn, result) => {
+  match (result) {
+    Ok(x) => Ok(fn(x)),
+    Err(e) => Err(e),
+  }
 }
 
-export let mapErr = (fn, v) => {
-    match (v) {
-        Err(e) => Err(fn(e)),
-        Ok(x)  => Ok(x)
-    }
+/**
+ * If the Result is `Err(x)`, returns `Err(fn(x))`. Returns the unmodified `Ok` otherwise.
+ *
+ * @param fn: The function to map
+ * @param result: The result to map on
+ * @returns `Err(fn(errValue))` if the result is of the `Err` variant, otherwise returns the unmodified `Ok`
+ */
+export let mapErr = (fn, result) => {
+  match (result) {
+    Err(e) => Err(fn(e)),
+    Ok(x) => Ok(x),
+  }
 }
 
-export let mapWithDefault = (fn, def, v) => {
-    match (v) {
-        Ok(a) => fn(a),
-        _     => def
-    }
+/**
+ * If the Result is `Ok(x)`, returns `fn(x)`. Returns the provided default otherwise.
+ *
+ * @param fn: The function to map
+ * @param def: The default value to return
+ * @param result: The result to map on
+ * @returns `Ok(fn(okValue))` if the result is of the `Ok` variant, otherwise returns the value given by the `def` param
+ */
+export let mapWithDefault = (fn, def, result) => {
+  match (result) {
+    Ok(a) => fn(a),
+    _ => def,
+  }
 }
 
-export let mapWithDefaultFn = (fnOk, fnErr, v) => {
-    match (v) {
-        Ok(a)  => fnOk(a),
-        Err(a) => fnErr(a)
-    }
+/**
+ * If the Result is `Ok(x)`, returns `fnOk(x)`. If the Result is `Err(y)`, returns `fnErr(y)`.
+ *
+ * @param fn: The function to map
+ * @param def: The default value to return
+ * @param result: The result to map on
+ * @returns `fnOk(okValue)` if the result is of the `Ok` variant, otherwise returns `fnErr(errValue)`
+ */
+export let mapWithDefaultFn = (fnOk, fnErr, result) => {
+  match (result) {
+    Ok(a) => fnOk(a),
+    Err(a) => fnErr(a),
+  }
 }
 
-export let or = (r1, r2) => {
-    match (r1) {
-        Ok(x) => r1,
-        _     => r2
-    }
+/**
+ * If the first Result is `Ok`, returns the first Result. Returns the second Result otherwise.
```

Comment:
I don't like this. How can we change it? I'd like to describe it as analogous to `a || b` but for Results.

Reply:
here is the description for js `||` on mdn ```
The logical OR (||) operator (logical disjunction) for a set of operands is true if and only if one or more of its operands is true. It is typically used with Boolean (logical) values. When it is, it returns a Boolean value. However, the || operator actually returns the value of one of the specified operands, so if this operator is used with non-Boolean values, it will return a non-Boolean value.
``` maybe it will help. maybe something like. ```
@returns The first given `Ok` Result, `None` otherwise
``` that is a little rough though so it probably needs some work still.

Reply:
```suggestion
 * Behaves like a logical OR (`||`) where the first Result is only returned if it is the `Ok` variant and falling back to the second Result in all other cases.
```
Thanks for finding that! What do you think about this workshop idea?

Reply:
I like it. maybe oscar or some one else has some ideas for improvments though.